3party/opening_hours/opening_hours_tests/osm_time_range_tests.cpp

```diff
@@ -0,0 +1,839 @@
+/*
+ The MIT License (MIT)
+
+ Copyright (c) 2015 Mail.Ru Group
+
+ Permission is hereby granted, free of charge, to any person obtaining a copy
+ of this software and associated documentation files (the "Software"), to deal
+ in the Software without restriction, including without limitation the rights
+ to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ copies of the Software, and to permit persons to whom the Software is
+ furnished to do so, subject to the following conditions:
+
+ The above copyright notice and this permission notice shall be included in
+ all copies or substantial portions of the Software.
+
+ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ THE SOFTWARE.
+*/
+
+#include <osm_time_range.hpp>
+#include <osm_parsers.hpp>
+
+#include <iostream>
+#include <sstream>
+#include <fstream>
+#include <map>
+#include <locale>
+
+#define BOOST_TEST_MODULE OpeningHours
+
+#if defined(__clang__)
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wshorten-64-to-32"
+#include <boost/test/included/unit_test.hpp>
+#pragma clang diagnostic pop
+#else
+#include <boost/test/included/unit_test.hpp>
+#endif
+
+#include <boost/spirit/include/qi.hpp>
+
+template <typename Char, typename Parser>
+bool test(Char const * in, Parser const & p, bool full_match = true)
+{
+  // we don't care about the result of the "what" function.
+  // we only care that all parsers have it:
+  boost::spirit::qi::what(p);
+
+  Char const * last = in;
+  while (*last)
+    last++;
+  return boost::spirit::qi::parse(in, last, p) && (!full_match || (in == last));
+}
+
+template <template <typename> class Parser, typename ParseResult, typename Char>
+std::basic_string<Char> ParseAndUnparse(Char const * input)
+{
+  // we don't care about the result of the "what" function.
+  // we only care that all parsers have it:
+
+  using boost::spirit::qi::phrase_parse;
+  using boost::spirit::standard_wide::space;
+
+  ParseResult parseResult;
+  std::basic_string<Char> const str(input);
+
+  Parser<decltype(begin(str))> p;
+  boost::spirit::qi::what(p);
+
+  auto first = begin(str);
+  auto const last = end(str);
+  auto parsed = boost::spirit::qi::phrase_parse(
+      first,
+      last,
+      p,
+      space,
+      parseResult);
+
+  if (!parsed || first != last)
```

Comment:
How about to change signature of ParseAndUnparse to:
```
bool ParseAndUnparse(Char const * input, std::basic_string<Char> & output);
```
and return true/false instead of that hacky ":CAN'T PARSE:" result?